File: lambda/create_page.py
"""
AWS Lambda function for creating new pages in Notion database
Handles POST requests to create case entries
"""
import logging
from typing import Dict, Any
from utils import (
    SecretsManager, 
    NotionClient, 
    NotionAPIError,
    validate_database_id,
    create_error_response,
    create_success_response,
    handle_preflight,
    parse_event_body
)

logger = logging.getLogger()
logger.setLevel(logging.INFO)

def validate_page_properties(properties: Dict[str, Any]) -> Dict[str, Any]:
    """
    Validate and format page properties for Notion API 2025-09-03
    """
    if not properties:
        raise ValueError("Properties are required")
    
    # Ensure Title property exists
    if 'Title' not in properties:
        raise ValueError("Title property is required")
    
    # Validate Title property format
    title_prop = properties['Title']
    if not isinstance(title_prop, dict) or 'title' not in title_prop:
        raise ValueError("Title property must have 'title' field")
    
    return properties

def lambda_handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler for creating new pages in Notion database
    
    Expected event structure:
    {
        "httpMethod": "POST",
        "body": "{
            \"databaseId\": \"40c4cef5c8cd4cb4891a35c3710df6e9\",
            \"properties\": {
                \"Title\": {
                    \"title\": [{\"text\": {\"content\": \"Case Title\"}}]
                },
                \"Case Number\": {
                    \"rich_text\": [{\"text\": {\"content\": \"CASE-2025-001\"}}]
                }
            }
        }"
    }
    """
    try:
        # Handle preflight CORS requests
        preflight_response = handle_preflight(event)
        if preflight_response:
            return preflight_response
        
        # Validate HTTP method
        http_method = event.get('httpMethod', '').upper()
        if http_method != 'POST':
            return create_error_response(f"Method {http_method} not allowed", 405)
        
        # Parse request body
        body = parse_event_body(event)
        
        # Extract and validate required fields
        database_id = body.get('databaseId')
        properties = body.get('properties')
        
        if not database_id:
            return create_error_response("Missing required field: databaseId", 400)
        
        if not properties:
            return create_error_response("Missing required field: properties", 400)
        
        # Validate database ID format
        clean_database_id = validate_database_id(database_id)
        
        # Validate properties
        validated_properties = validate_page_properties(properties)
        
        logger.info(f"Creating page in database: {clean_database_id}")
        
        # Initialize AWS services and Notion client
        secrets_manager = SecretsManager()
        notion_client = NotionClient(secrets_manager)
        
        # Check if this is a multi-source database and get data_source_id if needed
        data_source_id = notion_client.get_data_source_id(clean_database_id)
        
        # Prepare page data for Notion API
        page_data = {
            'parent': {
                'database_id': clean_database_id
            },
            'properties': validated_properties
        }
        
        # Add data_source_id if available (for multi-source databases in 2025-09-03)
        if data_source_id:
            page_data['parent']['data_source_id'] = data_source_id
            logger.info(f"Using data_source_id: {data_source_id}")
        
        # Create the page
        created_page = notion_client.create_page(page_data)
        
        # Extract relevant information for response
        response_data = {
            'id': created_page.get('id'),
            'url': created_page.get('url'),
            'created_time': created_page.get('created_time'),
            'properties': created_page.get('properties', {}),
            'api_version': '2025-09-03'
        }
        
        if data_source_id:
            response_data['data_source_id'] = data_source_id
        
        logger.info(f"Successfully created page: {response_data['id']}")
        return create_success_response(response_data, 201)
        
    except ValueError as e:
        # Handle validation errors
        return create_error_response(str(e), 400)
        
    except NotionAPIError as e:
        # Handle Notion API specific errors
        if e.status_code == 401:
            return create_error_response("Invalid Notion API token or insufficient permissions", 401)
        elif e.status_code == 404:
            return create_error_response("Database not found or integration lacks access", 404)
        elif e.status_code == 400:
            return create_error_response(f"Invalid request data: {e.message}", 400)
        elif e.status_code == 429:
            return create_error_response("Rate limit exceeded. Please try again later.", 429)
        else:
            return create_error_response(f"Notion API error: {e.message}", e.status_code)
    
    except Exception as e:
        # Handle unexpected errors
        logger.error(f"Unexpected error in create_page: {e}", exc_info=True)
        return create_error_response("Internal server error", 500)

# For testing purposes, allow direct invocation
if __name__ == "__main__":
    # Test event for local testing
    test_event = {
        "httpMethod": "POST",
        "body": '{"databaseId": "40c4cef5c8cd4cb4891a35c3710df6e9", "properties": {"Title": {"title": [{"text": {"content": "Test Case"}}]}}}'
    }
    
    result = lambda_handler(test_event, None)
    print(f"Test result: {result}")

File: api/createPage.js
import { createNotionClient, handlePreflight, sendError, sendSuccess, validateRequired } from './utils.js';

/**
 * Create a new page in Notion database
 * POST /api/createPage
 */
export default async function handler(req, res) {
  // Handle preflight requests
  if (handlePreflight(req, res)) {
    return;
  }

  try {
    // Only allow POST requests
    if (req.method !== 'POST') {
      return sendError(res, new Error(`Method ${req.method} not allowed`), 405);
    }

    // Validate required fields
    const { databaseId, properties } = req.body;
    if (!databaseId || !properties) {
      return sendError(res, new Error('Missing required fields: databaseId, properties'), 400);
    }

    // Validate database ID format
    if (!/^[a-f0-9]{32}$/i.test(databaseId.replace(/-/g, ''))) {
      return sendError(res, new Error('Invalid database ID format'), 400);
    }

    // Validate that Title property exists
    if (!properties.Title) {
      return sendError(res, new Error('Title property is required'), 400);
    }

    // Create Notion client
    const notion = createNotionClient();
    
    // Prepare basic page data structure
    const pageData = {
      parent: {
        database_id: databaseId
      },
      properties: properties
    };

    // Try to get data_source_id for multi-source databases (2025-09-03 API)
    try {
      const dbResponse = await notion.get(`/databases/${databaseId}`);
      const data_sources = dbResponse.data.data_sources;
      
      // Only add data_source_id if the database has data sources
      if (data_sources && data_sources.length > 0) {
        pageData.parent.data_source_id = data_sources[0].id;
        console.log(`Using data_source_id: ${data_sources[0].id} for multi-source database`);
      }
    } catch (dbError) {
      // If we can't fetch database info, continue without data_source_id
      // This handles cases where the database might not support the data_sources endpoint
      console.warn('Could not fetch database data sources, proceeding with standard database_id:', dbError.message);
    }

    const response = await notion.post('/pages', pageData);

    // Prepare response data with additional metadata
    const responseData = {
      id: response.data.id,
      url: response.data.url,
      created_time: response.data.created_time,
      properties: response.data.properties,
      api_version: '2025-09-03'
    };

    // Include data_source_id in response if it was used
    if (pageData.parent.data_source_id) {
      responseData.data_source_id = pageData.parent.data_source_id;
    }

    // Return success with created page info
    sendSuccess(res, responseData, 201);

  } catch (error) {
    // Handle Notion API errors specifically
    if (error.response) {
      const status = error.response.status;
      const notionError = error.response.data;
      
      if (status === 401) {
        return sendError(res, new Error('Invalid Notion API key'), 401);
      } else if (status === 404) {
        return sendError(res, new Error('Database not found or integration lacks access'), 404);
      } else if (status === 400) {
        return sendError(res, new Error(notionError.message || 'Invalid request data'), 400);
      } else if (status === 429) {
        return sendError(res, new Error('Rate limit exceeded. Please try again later.'), 429);
      } else {
        return sendError(res, new Error(notionError.message || 'Notion API error'), status);
      }
    }
    
    // Handle other errors
    sendError(res, error);
  }
}
